paper/src/main/java/com/denizenscript/denizen/paper/properties/PaperWorldExtensions.java

```diff
@@ -23,6 +29,81 @@ public static void register() {
             return new ElementTag(world.getWorld().getNoTickViewDistance());
         });
 
+        if (NMSHandler.getVersion().isAtLeast(NMSVersion.v1_20)) {
+
+            // <--[tag]
+            // @attribute <WorldTag.gateway_count>
+            // @returns ElementTag(Number)
+            // @description
+            // Returns the number of end gateway portals.
+            // Only works in end worlds.
+            // -->
+            WorldTag.tagProcessor.registerTag(ElementTag.class, "gateway_count", (attribute, object) -> {
+                DragonBattle battle = object.getWorld().getEnderDragonBattle();
+                if (battle == null) {
+                    attribute.echoError("Provided world is not an end world!");
+                    return null;
+                }
+                return new ElementTag(battle.getGatewayCount());
+            });
+
+            // <--[tag]
+            // @attribute <WorldTag.healing_crystals>
+            // @returns ListTag(EntityTag)
+            // @description
+            // Returns a ListTag of the healing crystals located on top of the obsidian towers.
+            // Only works in end worlds.
+            // -->
+            WorldTag.tagProcessor.registerTag(ListTag.class, "healing_crystals", (attribute, object) -> {
+                DragonBattle battle = object.getWorld().getEnderDragonBattle();
+                if (battle == null) {
+                    attribute.echoError("Provided world is not an end world!");
+                    return null;
+                }
+                return new ListTag(battle.getHealingCrystals(), EntityTag::new);
+            });
+
+            // <--[tag]
+            // @attribute <WorldTag.healing_crystals>
+            // @returns ListTag(EntityTag)
+            // @description
+            // Returns a ListTag of the respawn crystals located at the end exit portal.
+            // Only works in end worlds.
+            // -->
+            WorldTag.tagProcessor.registerTag(ListTag.class, "respawn_crystals", (attribute, object) -> {
+                DragonBattle battle = object.getWorld().getEnderDragonBattle();
+                if (battle == null) {
+                    attribute.echoError("Provided world is not an end world!");
+                    return null;
+                }
+                return new ListTag(battle.getRespawnCrystals(), EntityTag::new);
+            });
+
+            // <--[mechanism]
+            // @object WorldTag
+            // @name spawn_gateway
+            // @input LocationTag
+            // @Plugin Paper
+            // @group paper
+            // @description
+            // If no location is specified, tries to spawn a new end gateway using default game mechanics. Otherwise, spawns a new end gateway portal at the specified location.

```

Comment:
Nitpick, but meta is usually the other way around (`spawns gateway at specified location, specify no input to spawn vanilla one`)